### ISSUE TITLE
Unable to deploy Vapor 2.0 beta in Heroku

### ISSUE THREAD

**Issue by skyweb07** [vapor/vapor#954]
Hi guys, I'm trying to deploy the new Vapor beta to Heroku after upgrading my whole app to Vapor 2.0 but it's not resolving the dependencies, it's actually working on my local mac but when I push to remote I can't make it work. I think it's because heroku needs to install some new dependencies that are not on the basic build pack that you provide but the same happens in travis when I try to run tests, this is the error that heroku show me

```
Resolving https://github.com/vapor/fluent-provider.git at 1.0.0-beta.3
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install cmysql using your system-packager:
    apt-get install cmysql
note: you may be able to install cmysql using your system-packager:
    apt-get install cmysql
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install cmysql using your system-packager:
    apt-get install cmysql
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install cmysql using your system-packager:
    apt-get install cmysql
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install cmysql using your system-packager:
    apt-get install cmysql
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install cmysql using your system-packager:
    apt-get install cmysql
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install cmysql using your system-packager:
    apt-get install cmysql
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install cmysql using your system-packager:
    apt-get install cmysql
Compile Swift Module 'Debugging' (1 sources)
Compile Swift Module 'ServiceLocator' (3 sources)
Compile Swift Module 'Bits' (19 sources)
Compile Swift Module 'Jay' (21 sources)
Compile Swift Module 'PathIndexable' (2 sources)
Compile Swift Module 'libc' (1 sources)
Compile Swift Module 'Core' (18 sources)
Compile Swift Module 'Transport' (11 sources)
Compile Swift Module 'Random' (6 sources)
Compile Swift Module 'Node' (35 sources)
Compile Swift Module 'Console' (34 sources)
Compile Swift Module 'Sockets' (22 sources)
Compile Swift Module 'Crypto' (13 sources)
Compile Swift Module 'BCrypt' (9 sources)
Compile Swift Module 'TLS' (12 sources)
Compile CSQLite sqlite3.c
Compile Swift Module 'JSON' (9 sources)
Compile Swift Module 'Redis' (10 sources)
Compile Swift Module 'MySQL' (11 sources)
Compile Swift Module 'URI' (10 sources)
<module-includes>:1:10: note: in file included from <module-includes>:1:
#include "shim.h"
         ^
/tmp/build_5ca535539efcc4cf29dde278a98c136b/goRestart-restart-backend-app-c7cc656/.build/checkouts/cmysql.git-1318605283685182930/shim.h:4:10: error: 'mysql.h' file not found
#include <mysql.h>
         ^
/tmp/build_5ca535539efcc4cf29dde278a98c136b/goRestart-restart-backend-app-c7cc656/.build/checkouts/mysql.git-4167185222669414350/Sources/MySQL/Database.swift:1:8: error: could not build Objective-C module 'CMySQL'
import CMySQL
       ^
Compile Swift Module 'SMTP' (21 sources)
<unknown>:0: error: build had 1 command failures
swift-build: error: exit(1): /app/.swiftenv/versions/3.1/usr/bin/swift-build-tool -f /tmp/build_5ca535539efcc4cf29dde278a98c136b/goRestart-restart-backend-app-c7cc656/.build/release.yaml
 !     Push rejected, failed to compile Swift app.
 !     Push failed
``` 

If I'm not wrong this is because Heroku can't download cmysql dependency so it can't not import the module, could you guys help me out with this?

Thanks in advance!!!

**Comment by tanner0101** [comment]
You'll need to run:

```sh
eval "$(curl -sL https://apt.vapor.sh)"
sudo apt-get install ctls cmysql
```

**Comment by vzsg** [comment]
I think it's a little more complex than this.

On Heroku, system-wide apt-get is not available, neither is sudo.

There's an official [buildpack](https://github.com/heroku/heroku-buildpack-apt) that can be added as a secondary, for pulling in .deb packages with APT and "installing" them manually. That could provide the scaffolding for Heroku support in v2, but the custom repo needs some extra though and experimentation.

I think listing all the .deb packages with direct links should work.

**Comment by skyweb07** [comment]
thanks, I'll try to set the `Aptfile` dependencies to see if work, I'll let you know how this goes :)

**Comment by skyweb07** [comment]
@vzsg I've tried but the buildpack itself doesn't support adding external repos so I can't add the `deb https://repo.vapor.codes/apt trusty main` into so it can't find those dependencies, do you have any idea of how can this could be resolved? or maybe getting the deb url for those packages I think it could work

**Comment by skyweb07** [comment]
Even after using the `https://elements.heroku.com/buildpacks/heroku/heroku-buildpack-apt` pointing with the final deb files
`
https://repo.vapor.codes/apt/pool/trusty/main/c/cmysql/cmysql_1.0.0_all.deb
https://repo.vapor.codes/apt/pool/trusty/main/c/ctls/ctls_1.0.0_all.deb
`
still throw me the same errors that can't find those dependencies

**Comment by vzsg** [comment]
I haven't yet figured out how to solve the problem, but I'm logging what I found in the last iterations.

With this [Aptfile](https://github.com/vapor/vapor/files/905085/Aptfile.txt) (cmysql, ctls and their dependencies manually added from the control files), the apt buildpack installs the following files under `~/.apt`: 
[files.txt](https://github.com/vapor/vapor/files/905092/files.txt) and sets the following environment variables: 
[env.txt](https://github.com/vapor/vapor/files/905100/env.txt)

However, I couldn't get Swift to pick them up in any way.

**Comment by valeriomazzeo** [comment]
```
-----> Swift app detected
Cloning into 'swiftenv'...
Swift 3 Heroku Installer
????  Version: 3.1
????  Operating System: ubuntu1404
???? Installing Swiftenv
Cloning into '/app/.swiftenv'...
???? Installing Swift
Downloading https://swift.org/builds/swift-3.1-release/ubuntu1604/swift-3.1-RELEASE/swift-3.1-RELEASE-ubuntu16.04.tar.gz
/tmp/swiftenv-3.1- /tmp/build_b00b69df2f3aa40a8d3d69d672b3d8a2
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
/tmp/build_b00b69df2f3aa40a8d3d69d672b3d8a2
3.1 has been installed.
???  Done
precompile
-----> Building Package ... this will take a while
Fetching https://github.com/vapor/vapor.git
Fetching https://github.com/vapor/engine.git
Fetching https://github.com/vapor/console.git
Fetching https://github.com/vapor/json.git
Fetching https://github.com/vapor/routing.git
Fetching https://github.com/vapor/multipart.git
Fetching https://github.com/vapor/bcrypt.git
Fetching https://github.com/vapor/crypto.git
Fetching https://github.com/vapor/tls.git
Fetching https://github.com/vapor/random.git
Fetching https://github.com/vapor/ctls.git
Fetching https://github.com/vapor/core.git
Fetching https://github.com/vapor/bits.git
Fetching https://github.com/vapor/debugging.git
Fetching https://github.com/vapor/socks.git
Fetching https://github.com/vapor/node.git
Fetching https://github.com/DanToml/Jay.git
Fetching https://github.com/vapor/path-indexable.git
Cloning https://github.com/vapor/routing.git
Resolving https://github.com/vapor/routing.git at 2.0.0-beta.1
Cloning https://github.com/vapor/json.git
Resolving https://github.com/vapor/json.git at 2.0.0-beta.1
Cloning https://github.com/vapor/core.git
Resolving https://github.com/vapor/core.git at 2.0.0-beta1
Cloning https://github.com/vapor/vapor.git
Resolving https://github.com/vapor/vapor.git at 2.0.0-beta.1
Cloning https://github.com/vapor/socks.git
Resolving https://github.com/vapor/socks.git at 2.0.0-beta.1
Cloning https://github.com/vapor/path-indexable.git
Resolving https://github.com/vapor/path-indexable.git at 2.0.0-beta.1
Cloning https://github.com/vapor/bits.git
Resolving https://github.com/vapor/bits.git at 1.0.0-beta1
Cloning https://github.com/vapor/node.git
Resolving https://github.com/vapor/node.git at 2.0.0-beta.1
Cloning https://github.com/DanToml/Jay.git
Resolving https://github.com/DanToml/Jay.git at 1.0.1
Cloning https://github.com/vapor/random.git
Resolving https://github.com/vapor/random.git at 1.0.0-beta.1
Cloning https://github.com/vapor/tls.git
Resolving https://github.com/vapor/tls.git at 2.0.0-beta.1
Cloning https://github.com/vapor/bcrypt.git
Resolving https://github.com/vapor/bcrypt.git at 1.0.0-beta.1
Cloning https://github.com/vapor/crypto.git
Resolving https://github.com/vapor/crypto.git at 2.0.0-beta.1
Cloning https://github.com/vapor/debugging.git
Resolving https://github.com/vapor/debugging.git at 1.0.0-beta1
Cloning https://github.com/vapor/engine.git
Resolving https://github.com/vapor/engine.git at 2.0.0-beta.1
Cloning https://github.com/vapor/console.git
Resolving https://github.com/vapor/console.git at 2.0.0-beta.2
Cloning https://github.com/vapor/multipart.git
Resolving https://github.com/vapor/multipart.git at 2.0.0-beta.1
Cloning https://github.com/vapor/ctls.git
Resolving https://github.com/vapor/ctls.git at 1.0.0-beta.4
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
note: you may be able to install ctls using your system-packager:
    apt-get install ctls
Compile Swift Module 'Debugging' (1 sources)
Compile Swift Module 'Bits' (19 sources)
Compile Swift Module 'PathIndexable' (2 sources)
Compile Swift Module 'Jay' (21 sources)
Compile Swift Module 'libc' (1 sources)
Compile Swift Module 'Core' (18 sources)
Compile Swift Module 'Random' (5 sources)
Compile Swift Module 'Node' (35 sources)
Compile Swift Module 'Console' (34 sources)
Compile Swift Module 'Transport' (11 sources)
Compile Swift Module 'Sockets' (22 sources)
Compile Swift Module 'Crypto' (13 sources)
Compile Swift Module 'BCrypt' (9 sources)
Compile Swift Module 'TLS' (12 sources)
Compile Swift Module 'JSON' (9 sources)
Compile Swift Module 'SMTP' (21 sources)
Compile Swift Module 'URI' (10 sources)
Compile Swift Module 'HTTP' (37 sources)
Compile Swift Module 'WebSockets' (14 sources)
Compile Swift Module 'Cookies' (11 sources)
Compile Swift Module 'Multipart' (6 sources)
Compile Swift Module 'Branches' (3 sources)
Compile Swift Module 'Routing' (8 sources)
Compile Swift Module 'FormData' (4 sources)
Compile Swift Module 'TypeSafeRouting' (3 sources)
Compile Swift Module 'Settings' (10 sources)
Compile Swift Module 'Cache' (2 sources)
Compile Swift Module 'Sessions' (7 sources)
Compile Swift Module 'Vapor' (76 sources)
Compile Swift Module 'Testing' (6 sources)
Compile Swift Module 'MyModule' (26 sources)
Compile Swift Module 'App' (1 sources)
Linking ./.build/release/App
-----> Installing dynamic libraries
-----> Installing binaries
-----> Discovering process types
       Procfile declares types -> web
-----> Compressing...
       Done: 65.6M
-----> Launching...
       Released v48
       https://myapp.herokuapp.com/ deployed to Heroku
```

Mine is working, but I am not using any database for my app...

**Comment by vzsg** [comment]
> Mine is working, but I am not using any database for my app...

That's great info! So only MySQL is problematic, but without is, apps should work on Heroku even with v2.

**Comment by skyweb07** [comment]
mmm I'll give it a try without MySQL

**Comment by skyweb07** [comment]
Haleluya!, it was MySQL dependency the one that was making the app not compile at all, so after removing that dependency the deploy was successfully, thanks guys!!

**Comment by vzsg** [comment]
Oh, I was under the impression that MySQL is required for your app 😛

**Comment by valeriomazzeo** [comment]
@vzsg shouldn't be possible to update the build pack to include that dependency?

**Comment by vzsg** [comment]
If you scroll back, you can see that I was experimenting with it, but in the end I did not succeed in the end. Even with Heroku's limitations, it should be possible, but getting Swift to find the headers is problematic.

**Comment by BrettRToomey** [comment]
@vzsg Any updates?

**Comment by vzsg** [comment]
Nope. I'll wait for #980 to be merged (which fixes a blocking bug) and give it a try again.

**Comment by tanner0101** [comment]
#980 has been merged and tagged as 2.0.0-beta.19

**Comment by vzsg** [comment]
Okay, I tested the same setup, still CMySQL is still broken.
But I'm closing this, as the original issue was resolved by removing the DB dependency, and will open an issue on vapor/cmysql instead.
